File: day09/part2.rb
class DoubleLinkedCircularList
  attr_reader :length

  class Node
    attr_accessor :previous
    attr_accessor :next
    attr_reader   :value

    def initialize(_value)
      @previous = nil
      @next = nil
      @value = _value
    end

    def to_s
      @value.to_s
    end
  end

  def initialize
    @initial = nil
    @current = nil
    @length = 0
  end

  def step(count)
    if @current
      count.abs.times do
        @current = if count < 0
                     @current.previous
                   else
                     @current.next
        end
      end
    end
    self
  end

  def insert(value)
    new_node = Node.new(value)
    if @current.nil?
      new_node.previous = new_node
      new_node.next = new_node
    else
      new_node.previous = @current
      new_node.next = @current.next
    end
    new_node.previous.next = new_node
    new_node.next.previous = new_node
    @current = new_node
    @initial = new_node if @initial.nil?
    @length += 1
    self
  end

  def delete
    return self if @current.nil?

    previous_node = @current.previous
    previous_node.next = @current.next
    previous_node.next.previous = previous_node
    if @current == @current.next
      @current = nil
      @initial = nil
    else
      @initial = @current.next if @initial == @current
      @current = @current.previous
    end
    @length -= 1
    self
  end

  def read
    if @current.nil?
      nil
    else
      @current.value
    end
  end

  def to_s
    str = '['
    node = @initial
    unless node.nil?
      loop do
        str += node.to_s
        str += '*' if node == @current
        node = node.next
        break if node == @initial

        str += ','
      end
    end
    str += ']'
    str
  end
end

lines = File.readlines('input')
lines.each do |_line|
  player_count = _line.match(/(\d+) players/).to_a[1].to_i
  last_marble = _line.match(/(\d+) points/).to_a[1].to_i

  marbles = DoubleLinkedCircularList.new
  marbles.insert(0)
  scores = Hash.new(0)
  (1..last_marble * 100).each do |_marble|
    player = (_marble - 1) % player_count
    if _marble % 23 == 0
      marbles.step(-7)
      scores[player] += _marble + marbles.read
      marbles.delete
      marbles.step(1)
    else
      marbles.step(1)
      marbles.insert(_marble)
    end
    # puts "#{player + 1} #{marbles}"
  end

  expected = _line.match(/high score is (\d+)/).to_a[1].to_i if _line.include? 'high score is'
  if expected
    score = scores.values.max
    puts "score: #{score}, expected: #{expected}"
    # exit
  end

  puts scores.values.max
end

File: day09/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type doublyLinkedCircularListNode struct {
	previous *doublyLinkedCircularListNode
	next *doublyLinkedCircularListNode
	value  interface{}
}

type DoublyLinkedCircularList struct {
	length int
	head *doublyLinkedCircularListNode
}

func NewDoublyLinkedCircularList() *DoublyLinkedCircularList {
	return &DoublyLinkedCircularList{
		length: 0,
		head: nil,
	}
}

func (self *DoublyLinkedCircularList) Rotate(steps int) bool {
	if self.head == nil {
		return false
	}
	count := steps
	if count<0 {
		count *= -1
	}
	for i := 0; i < count; i++ {
		if steps < 0 {
			self.head = self.head.previous
		} else {
			self.head = self.head.next
		}
	}
	return true
}

func (self *DoublyLinkedCircularList) Append(value interface{}) bool {
	node := &doublyLinkedCircularListNode{
		previous: nil,
		next: nil,
		value: value,
	}
	if self.head == nil {
      node.previous = node
      node.next = node
	} else {
      node.previous = self.head
      node.next = self.head.next
	}
	node.previous.next = node
    node.next.previous = node
    self.head = node
    self.length++
	return true
}

func (self *DoublyLinkedCircularList) Remove() bool {
	if self.head == nil {
		return false
	}
    node := self.head.previous
    node.next = self.head.next
    node.next.previous = node
    if self.head == self.head.next {
    	self.head = nil
	} else {
        self.head = self.head.previous
	}
    self.length--
	return true
}

func (self *DoublyLinkedCircularList) Read() interface{} {
	if self.head == nil {
		return nil
	}
	return self.head.value
}

func (self *DoublyLinkedCircularList) Length() interface{} {
	return self.length
}

func (self *DoublyLinkedCircularList) ToString() string {
	str := "["
	if self.head != nil {
		node := self.head
		for {
			node = node.next
			str += fmt.Sprintf("%v",node.value)
			if node == self.head {
				break
			}
			str += ","
		}
	}
	str += "]"
	return str
}

func main() {
	f, _ := os.Open("input")
	defer f.Close()
	s := bufio.NewScanner(f)
	for s.Scan() {
		line := s.Text()
		re1 := regexp.MustCompile("(\\d+) players")
		playerCount,_ := strconv.Atoi(re1.FindStringSubmatch(line)[1])
		re2 := regexp.MustCompile("(\\d+) points")
		lastMarble,_:= strconv.Atoi(re2.FindStringSubmatch(line)[1])

		marbles := NewDoublyLinkedCircularList()
		marbles.Append(0)
		scores := map[int]int{}
		for marble := 1; marble <= lastMarble*100; marble++ {
			player := (marble - 1) % playerCount
			if marble % 23 == 0 {
				marbles.Rotate(-7)
				if _, ok := scores[player]; !ok {
					scores[player] = 0	
				}				
				scores[player] += marble + marbles.Read().(int)
				marbles.Remove()
				marbles.Rotate(1)
			} else {
				marbles.Rotate(1)
				marbles.Append(marble)
			}
		}

		score := 0
		for _, v := range scores { 
			if v>score {
				score = v
			}	
		}

		if strings.Contains(line, "high score is") {
			re3 := regexp.MustCompile("high score is (\\d+)")
			expected,_:= strconv.Atoi(re3.FindStringSubmatch(line)[1])
			fmt.Printf("score: %v, expected: %v\n", score, expected)
		}

		fmt.Printf("%v\n", score)
	}

}